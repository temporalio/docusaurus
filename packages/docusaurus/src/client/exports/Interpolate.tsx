/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, {ReactNode} from 'react';

/*
Minimal implementation of a React interpolate component.
We don't ship a markdown parser nor a feature-complete i18n library on purpose.
More details here: https://github.com/facebook/docusaurus/pull/4295
*/

const ValueRegexp = /{\w+}/g;
const ValueFoundMarker = '{}'; // does not care much

// TODO use TS template literal feature to make values typesafe!
// (requires upgrading TS first)
// eslint-disable-next-line @typescript-eslint/no-unused-vars
type ExtractInterpolatePlaceholders<Str extends string> = string;

type InterpolateValues<Str extends string, Value extends ReactNode> = Record<
  ExtractInterpolatePlaceholders<Str>,
  Value
>;

// TS function overload: if all the values are plain strings, then interpolate returns a simple string
export function interpolate<Str extends string>(
  text: Str,
  values?: InterpolateValues<Str, string | number>,
): string;

// If values contain any ReactNode, then the return is a ReactNode
export function interpolate<Str extends string, Value extends ReactNode>(
  text: Str,
  values?: InterpolateValues<Str, Value>,
): ReactNode;

export function interpolate<Str extends string, Value extends ReactNode>(
  text: Str,
  values?: InterpolateValues<Str, Value>,
): ReactNode {
  const elements: (Value | string)[] = [];

  const processedText = text.replace(ValueRegexp, (match: string) => {
    // remove {{ and }} around the placeholder
    const key = match.substr(
      1,
      match.length - 2,
    ) as ExtractInterpolatePlaceholders<Str>;

    const value = values?.[key];

    if (typeof value !== 'undefined') {
      const element = React.isValidElement(value)
        ? value
        : // For non-React elements: basic primitive->string conversion
          String(value);
      elements.push(element);
      return ValueFoundMarker;
    } else {
      return match; // no match? add warning?
    }
  });

  // No interpolation to be done: just return the text
  if (elements.length === 0) {
    return text;
  }
  // Basic string interpolation: returns interpolated string
  else if (elements.every((el) => typeof el === 'string')) {
    return processedText
      .split(ValueFoundMarker)
      .reduce<string>((str, value, index) => {
        return str.concat(value).concat((elements[index] as string) ?? '');
      }, '');
  }
  // JSX interpolation: returns ReactNode
  else {
    return processedText
      .split(ValueFoundMarker)
      .reduce<ReactNode[]>((array, value, index) => {
        return [
          ...array,
          <React.Fragment key={index}>
            {value}
            {elements[index]}
          </React.Fragment>,
        ];
      }, []);
  }
}

export type InterpolateProps<Str extends string> = {
  children: Str;
  values?: InterpolateValues<Str, ReactNode>;
};

export default function Interpolate<Str extends string>({
  children,
  values,
}: InterpolateProps<Str>) {
  return interpolate(children, values);
}
