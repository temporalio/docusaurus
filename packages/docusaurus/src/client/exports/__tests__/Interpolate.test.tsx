/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import {interpolate} from '../Interpolate';

describe('Interpolate', () => {
  test('without placeholders', () => {
    const text = 'Hello how are you?';
    expect(interpolate(text)).toEqual(text);
  });

  test('placeholders with string values', () => {
    const text = 'Hello {name} how are you {day}?';
    const values = {name: 'Sébastien', day: 'today'};
    expect(interpolate(text, values)).toMatchInlineSnapshot(
      `"Hello Sébastien how are you today?"`,
    );
  });

  test('placeholders with string values', () => {
    const text = '{number} {string} {object} {array}';
    const values = {
      number: 42,
      string: 'Hello',
      object: {hello: 'world'},
      array: ['Hello'],
    };
    // Do we need to improve the JS type -> String conversion logic here?
    expect(interpolate(text, values)).toMatchInlineSnapshot(
      `"42 Hello [object Object] Hello"`,
    );
  });

  test('placeholders with falsy values', () => {
    const text = '{number} {string} {boolean}';
    const values = {
      number: 0,
      string: '',
      boolean: false,
    };
    // Do we need to improve the JS type -> String conversion logic here?
    expect(interpolate(text, values)).toMatchInlineSnapshot(`"0  false"`);
  });

  test('placeholders with string values mismatch', () => {
    // Should we emit warnings in such case?
    const text = 'Hello {name} how are you {unprovidedValue}?';
    const values = {name: 'Sébastien', extraValue: 'today'};
    expect(interpolate(text, values)).toMatchInlineSnapshot(
      `"Hello Sébastien how are you {unprovidedValue}?"`,
    );
  });

  test('placeholders with values not provided', () => {
    // Should we emit warnings in such case?
    const text = 'Hello {name} how are you {day}?';
    expect(interpolate(text)).toEqual(text);
    expect(interpolate(text, {})).toEqual(text);
  });

  test('placeholders with JSX values', () => {
    const text = 'Hello {name} how are you {day}?';
    const values = {name: <b>Sébastien</b>, day: <span>today</span>};
    expect(interpolate(text, values)).toMatchSnapshot();
  });

  test('placeholders with mixed vales', () => {
    const text = 'Hello {name} how are you {day}?';
    const values = {name: 'Sébastien', day: <span>today</span>};
    expect(interpolate(text, values)).toMatchSnapshot();
  });

  test('acceptance test', () => {
    const text = 'Hello {name} how are you {day}? Another {unprovidedValue}!';
    const values = {
      name: 'Sébastien',
      day: <span>today</span>,
      extraUselessValue1: <div>test</div>,
      extraUselessValue2: 'hi',
    };
    expect(interpolate(text, values)).toMatchSnapshot();
  });
});
