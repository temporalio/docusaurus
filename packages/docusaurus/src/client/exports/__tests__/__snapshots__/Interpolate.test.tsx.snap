// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interpolate acceptance test 1`] = `
Array [
  <React.Fragment>
    Hello 
    Sébastien
  </React.Fragment>,
  <React.Fragment>
     how are you 
    <span>
      today
    </span>
  </React.Fragment>,
  <React.Fragment>
    ? Another {unprovidedValue}!
  </React.Fragment>,
]
`;

exports[`Interpolate placeholders with JSX values 1`] = `
Array [
  <React.Fragment>
    Hello 
    <b>
      Sébastien
    </b>
  </React.Fragment>,
  <React.Fragment>
     how are you 
    <span>
      today
    </span>
  </React.Fragment>,
  <React.Fragment>
    ?
  </React.Fragment>,
]
`;

exports[`Interpolate placeholders with mixed vales 1`] = `
Array [
  <React.Fragment>
    Hello 
    Sébastien
  </React.Fragment>,
  <React.Fragment>
     how are you 
    <span>
      today
    </span>
  </React.Fragment>,
  <React.Fragment>
    ?
  </React.Fragment>,
]
`;
